File: 2023/01/solution.rb
#!/usr/bin/env ruby

class Solution
  def self.test
    new(sample).run
  end

  def self.answer
    new(input).run
  end

  def self.input
    @input ||= File.readlines('input')
  end

  def self.sample
    %w(
      two1nine
      eightwothree
      abcone2threexyz
      xtwone3four
      4nineeightseven2
      zoneight234
      7pqrstsixteen
    )
  end

  attr_reader :result

  def initialize(input)
    @result = build(input)
  end

  def run
    # puts "sample: #{result}"
    puts "sum: #{result.sum}"
  end

  private

  def build(input)
    input.map do |line|
      "#{first(from_left(line))}#{first(from_right(line))}".to_i
    end
  end

  def from_left(line)
    line.gsub(/\d|#{dictionary.keys.join('|')}/) do |matched|
      parse(matched)
    end
  end

  def from_right(line)
    line.reverse.gsub(/\d|#{dictionary.keys.map(&:reverse).join('|')}/) do |matched|
      parse(matched.reverse)
    end
  end

  def first(string)
    string.gsub(/\D/, "")[0]
  end

  def parse(string)
    begin
      Integer(string)
      string
    rescue ArgumentError
      dictionary[string].to_i
    end
  end

  def dictionary
    {
      "eight" => 8,
      "one" => 1,
      "two" => 2,
      "three" => 3,
      "four" => 4,
      "five" => 5,
      "six" => 6,
      "seven" => 7,
      "nine" => 9,
      "zero" => 0,
    }
  end
end

Solution.test
Solution.answer

File: 2023/01/solution.py
#! usr/bin/python3

import re
from typing import List, Dict

def sample() -> List[str]:
    return [
        'two1nine',
        'eightwothree',
        'abcone2threexyz',
        'xtwone3four',
        '4nineeightseven2',
        'zoneight234',
        '7pqrstsixteen'
    ]

def input() -> List[str]:
    with open('input', 'r') as f:
        return f.readlines()

def check(test, result, expectation):
    print(f'{test} -- expect {result} to equal {expectation}: {result == expectation}')

class Solution:
    def __init__(self, lines):
        self.lines = lines
        self.mapper: Dict[str, int] = {
            "eight": 8,
            "one": 1,
            "two": 2,
            "three": 3,
            "four": 4,
            "five": 5,
            "six": 6,
            "seven": 7,
            "nine": 9,
            "zero": 0,
        }

    def answer(self) -> int:
        print(sum(self.build()))

    def build(self):
        result = []
        for line in  self.lines:
            combined = f"{self.from_left(line)}{self.from_right(line)}"
            result.append(int(combined))

        return result

    def from_left(self, line: str) -> str:
        regex = re.compile("|".join(list(self.mapper.keys()) + ["\\d"]))
        return [self.parse(m) for m in regex.findall(line)][0]

    def from_right(self, line: str) -> str:
        _mapper_keys_reversed = ["".join(reversed(k)) for k in self.mapper.keys()]
        regex = re.compile("|".join(_mapper_keys_reversed + ["\\d"]))
        return [self.parse("".join(reversed(m))) for m in regex.findall("".join(reversed(line)))][0]

    def parse(self, string: str) -> str:
        return int(string) if string.isdigit() else self.mapper[string]


# sol = Solution(sample())
# check("parse", sol.parse('1'), 1)
# check("parse", sol.parse('one'), 1)
# check("from_left", sol.from_left('zoneight234'), 1)
# check("from_right", sol.from_right('zoneight234'), 4)
# check("builds", sol.build(), [29, 83, 13, 24, 42, 14, 76 ])
Solution(sample()).answer()
Solution(input()).answer()
